### PR TITLE
fix(primitives): use a WHATWG compliant implementation of AbortController/AbortSignal

### PR DESCRIPTION
**Pull request by feugy**
### 📖 What's in there?

Fix https://github.com/vercel/next.js/issues/38658

- fix(primitives): use a WHATWG compliant implementation of AbortController/AbortSignal
- chore: uses jest watch typeahead plugin

### 🧪 How to test?

1. `pnpm clean && pnpm i && pnpm build`
2. `pnpm --filter primitives test`

If you want to try out the watcher plugin:
1. `pnpm --filter primitives test -- --watchAll`
2. then hit _p_ key to filter tests files, or _t_ key to filter tests names
3. use arrows + _enter_ to select the one you want, or hit _enter_ straight to run all matching files

### 🆙 Notes to reviewers

Our current implementation does have `abortController.signal` (it wouldn't work at all otherwise), but is lacking [`abortSignal.throwIfAborted()`](https://dom.spec.whatwg.org/#dom-abortsignal-throwifaborted) and [`AbortSignal.timeout()`](https://dom.spec.whatwg.org/#dom-abortsignal-timeout)

[abortcontroller-polyfill](https://github.com/mo/abortcontroller-polyfill/blob/master/src/abortcontroller.js#L46) and [yet-another-abortcontroller-polyfill](https://github.com/benlesh/abort-controller-polyfill/blob/master/index.js#L9) do not implement throwIfAborted() nor timeout()

[@flemist/abort-controller](https://github.com/NikolayMakhonin/abort-controller/blob/master/packages/abort-controller/README.md) is the closest, but does not have `timeout()` nor `abort()`. Rather than re-implementing from scratch, I'll used it and extended it.

~~The abort-controller file size has decreased from 20.1 KB to 6.39 KB~~
The abort-controller file size has decreased from 20.1 KB to 3.19 KB 🎉 

**Edit:** `@flemist/abort-controller` does not run on node@12 nor node@14, which we need to support.
I've inspired from it to provide a working, compliant implementation.

### CODE REVIEW

**Review by javivelasco — APPROVED** [review]

File: packages/primitives/package.json [hunk 1]
```diff
@@ -23,10 +23,10 @@
   ],
   "devDependencies": {
     "@edge-runtime/format": "workspace:1.1.0-beta.27",
+    "@flemist/abort-controller": "0.0.6",
```
Comment (Kikobeats) [inline packages/primitives/package.json:26]:
```suggestion
    "@flemist/abort-controller": "0.0.6",
```

**Comment by Kikobeats** [comment]
I like the jest plugin, but it looks like implies so many changes across all the packages.

What if we extract the jest [configuration into a file](https://jestjs.io/docs/configuration), so we keep it just in one place?

File: packages/primitives/package.json (hunk 1 above)
Comment (Kikobeats) [reply packages/primitives/package.json:26]: I think you don't need this anymore 😛

Comment (feugy) [reply packages/primitives/package.json]: Good catch, thank you!

File: packages/primitives/src/primitives/abort-controller.js [hunk 2]
```diff
@@ -1 +1,89 @@
...
+  if (typeof reason === 'undefined') {
+    reason = new DOMException('The operation was aborted.', 'AbortError')
+  }
+  if (signal.aborted) {
+    return
+  }
+
+  signal[kReason] = reason
+  signal[kAborted] = true
+  signal.dispatchEvent({ type: 'abort' }) // TODO: why can't we use `new Event('abort')` ??
```
[the remark this answers is not in the document]
Comment (feugy) [reply packages/primitives/src/primitives/abort-controller.js]: We actually can't, but I found out the issue.
It turned out we were exporting an empty `Event` constructor from an old version of  `event-target-shim`

I've updated the shim version, and used it. we're good now!

**Comment by feugy** [comment]
@Kikobeats we can not share the same file because of the way [jest is handling its rootDir](https://jestjs.io/docs/configuration#rootdir-string).
But used an easy trick to share 95% of the configuration, and let individual packages tweak what they need to.

**Pull request merged**